### PR TITLE
Start HUP signal handler earlier

### PR DESCRIPTION
**Pull request by miekg**
When prometheus starts up and is recovering its state it will not handle
SIGHUPs. If it receives those during this phase it will exit. The change
here makes prometheus ignore SIGHUPs until it is ready to handle them.

Note this is only done for SIGHUP because that signal is used for
trigger a config reload and a such something could already be sending
these signals as part of a config update.

### CODE REVIEW

**Comment by fabxc** [comment]
Good catch, thanks.

Could we turn the bool into a channel that blocks the goroutine up to the point we are able to handle `SIGHUP` so the signal is not lost?

**Comment by juliusv** [comment]
Also, `hupIgnore` is not protected by a mutex, but accessed from two goroutines. If it's still in the channel-based blocking solution, it needs to be protected somehow.

**Comment by miekg** [comment]
@juliusv ah, I though flipping a bool would be just fine, but we could do something atomic as well here.

**Comment by juliusv** [comment]
Yep, either use something from https://golang.org/pkg/sync/atomic/ or use a channel to communicate to the goroutine that we reached the point where we're ready to handle signals (better).

**Comment by miekg** [comment]
ack I will go with a channel based solution.

**Comment by juliusv** [comment]
So instead of the boolean `hupIgnore`, just block on reading from a channel in the signal handler (before trying to handle signals), and then `close()` the channel where you currently set the boolean to `true`.

**Comment by miekg** [comment]
PTAL.

File: main.go [hunk 1]
```diff
@@ -280,13 +293,8 @@ func (p *prometheus) Serve() {
...
 
 	// Wait for reload or termination signals.
-	hup := make(chan os.Signal)
-	signal.Notify(hup, syscall.SIGHUP)
-	go func() {
-		for range hup {
-			p.reloadConfig()
-		}
-	}()
+	block <- true // Unblock SIGHUP handler.
```
Comment (juliusv) [inline main.go]: Just the `close()` should be enough, no writing into the channel needed.

**Comment by fabxc** [comment]
Could you squash please and remove the `web/blob/files.go`? This is created on `make` due to a problem fixed in #793.

File: main.go [hunk 2]
```diff
@@ -247,6 +247,19 @@ func (p *prometheus) reloadConfig() bool {
 // down. The method installs an interrupt handler, allowing to trigger a
 // shutdown by sending SIGTERM to the process.
 func (p *prometheus) Serve() {
+	// Wait for reload or termination signals. Start the handler for SIGHUP as
+	// early as possible, but ignore it until we are ready to handle reloading
+	// our config.
+	hup := make(chan os.Signal)
+	block := make(chan bool)
```
Comment (juliusv) [inline main.go]: Could you call this `hupReady` or something that makes it easier to understand the purpose?

**Comment by juliusv** [comment]
:+1: Nice, thanks a bunch!

**Comment by miekg** [comment]
\o/

**Pull request merged**